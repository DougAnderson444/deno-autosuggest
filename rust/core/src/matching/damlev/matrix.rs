use std::fmt;


pub struct DistMatrix {
    size: usize,
    raw: Vec<f64>,
}


impl DistMatrix {
    pub fn new(size: usize) -> Self {
        let raw = vec![0.0; size * size];
        let mut matrix = Self { size, raw };
        matrix.init();
        matrix
    }

    pub fn grow(&mut self, size: usize) {
        if size <= self.size {
            return;
        }
        let size = size + size / 2;
        self.raw.resize(size * size, 0.0);
        self.size = size;
        self.init();
    }

    pub fn init(&mut self) {
        if self.size == 0 { return; }
        unsafe {
            for i in 0..self.size {
                self.set_unchecked(i, 0, self.size as f64);
                self.set_unchecked(0, i, self.size as f64);
            }
            for i in 1..self.size {
                self.set_unchecked(i, 1, (i - 1) as f64);
                self.set_unchecked(1, i, (i - 1) as f64);
            }
        }
    }

    #[inline]
    pub unsafe fn get_unchecked(&self, i: usize, j: usize) -> f64 {
        *self.raw.get_unchecked(i * self.size + j)
    }

    #[inline]
    pub unsafe fn set_unchecked(&mut self, i: usize, j: usize, val: f64) {
        *self.raw.get_unchecked_mut(i * self.size + j) = val;
    }

    #[inline]
    pub fn get(&self, i: usize, j: usize) -> f64 {
        self.raw[i * self.size + j]
    }
}


impl fmt::Debug for DistMatrix {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut imax = self.size;
        let mut jmax = self.size;
        while imax > 0 && self.get(imax - 1, 2) <= std::f64::EPSILON { imax -= 1; }
        while jmax > 0 && self.get(2, jmax - 1) <= std::f64::EPSILON { jmax -= 1; }

        write!(f, "DistMatrix:\n")?;
        for i in 2 .. imax {
            for j in 2 .. jmax {
                write!(f, "{:2} ", self.get(i, j))?;
            }
            write!(f, "\n")?;
        }
        Ok(())
    }
}