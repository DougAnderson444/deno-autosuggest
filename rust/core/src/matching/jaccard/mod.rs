use std::cell::RefCell;
use std::hash::Hash;
use std::collections::HashSet;


const DEFAULT_CAPACITY: usize = 20;


pub struct Jaccard<T: PartialEq + Copy + Ord + Hash> {
    set1: RefCell<HashSet<T>>,
    set2: RefCell<HashSet<T>>,
}


impl<T: PartialEq + Copy + Ord + Hash> Jaccard<T> {
    pub fn new() -> Self {
        Self {
            set1: RefCell::new(HashSet::with_capacity(DEFAULT_CAPACITY)),
            set2: RefCell::new(HashSet::with_capacity(DEFAULT_CAPACITY)),
        }
    }

    pub fn similarity(&self, slice1: &[T], slice2: &[T]) -> f64 {
        match (slice1.len(), slice2.len()) {
            (0, 0) => return 1.0,
            (0, _) => return 0.0,
            (_, 0) => return 0.0,
            (_, _) => { },
        }

        let set1 = &mut *self.set1.borrow_mut();
        let set2 = &mut *self.set2.borrow_mut();

        set1.clear();
        set2.clear();
        for &item1 in slice1 { set1.insert(item1); }
        for &item2 in slice2 { set2.insert(item2); }

        let count_union        = set1.union(set2).count();
        let count_intersection = set1.intersection(set2).count();

        count_intersection as f64 / count_union as f64
    }

    pub fn rel_dist(&self, slice1: &[T], slice2: &[T]) -> f64 {
        1.0 - self.similarity(slice1, slice2)
    }
}


#[cfg(test)]
mod tests {
    use super::Jaccard;


    fn round2(x: f64) -> f64 {
        (x * 100.0).round() / 100.0
    }


    #[test]
    fn empty_both() {
        let jaccard = Jaccard::new();
        let s1: &[usize] = &[];
        let s2: &[usize] = &[];
        assert_eq!(round2(jaccard.rel_dist(s1, s2)), 0.0);
    }


    #[test]
    fn empty_one() {
        let jaccard = Jaccard::new();
        let sample = [
            (1.00, vec![], vec![1]),
            (1.00, vec![], vec![1, 2]),
            (1.00, vec![], vec![1, 2, 3]),
            (1.00, vec![], vec![1, 2, 3, 4]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn equal() {
        let jaccard = Jaccard::new();
        let sample = [
            vec![1],
            vec![1, 2],
            vec![1, 2, 3],
            vec![1, 2, 3, 4],
        ];
        for s in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s, s)), 0.0);
        }
    }


    #[test]
    fn different() {
        let jaccard = Jaccard::new();
        let sample = [
            (1.00, vec![1],          vec![5]),
            (1.00, vec![1, 2],       vec![5, 6]),
            (1.00, vec![1, 2, 3],    vec![5, 6, 7]),
            (1.00, vec![1, 2, 3, 4], vec![5, 6, 7, 8]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn partial() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.25, vec![1, 2, 3, 4], vec![1, 2, 3]),
            (0.50, vec![1, 2, 3, 4], vec![1, 2]),
            (0.75, vec![1, 2, 3, 4], vec![1]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn permutated() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.00, vec![1, 2, 3, 4], vec![3, 1, 2, 4]),
            (0.00, vec![1, 2, 3, 4], vec![4, 2, 3, 1]),
            (0.00, vec![1, 2, 3, 4], vec![1, 4, 2 ,3]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn repeated() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.75, vec![1, 2, 3, 4], vec![1, 1, 1, 1]),
            (0.50, vec![1, 2, 3, 4], vec![2, 1, 2, 1]),
            (0.25, vec![1, 2, 3, 4], vec![3, 2, 1, 3]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }


    #[test]
    fn complex() {
        let jaccard = Jaccard::new();
        let sample = [
            (0.40, vec![1, 2, 3, 4], vec![0, 1, 2, 4]),
            (0.40, vec![1, 2, 3, 4], vec![0, 2, 3, 1]),
            (0.50, vec![1, 2, 3, 4], vec![0, 4, 2 ,3, 5]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(round2(jaccard.rel_dist(s1, s2)), *d);
        }
    }
}