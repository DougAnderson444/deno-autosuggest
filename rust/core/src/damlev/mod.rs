mod matrix;
mod utils;

use std::cmp::max;
use std::collections::BTreeMap;
use std::cell::RefCell;
use utils::common_affix_sizes;
use matrix::DistMatrix;


const DEFAULT_CAPACITY: usize = 20;


pub struct DamerauLevenshtein<T: PartialEq + Copy + Ord> {
    pub dists: RefCell<DistMatrix>,
    last_i1: RefCell<BTreeMap<T, usize>>,
}


impl<T: PartialEq + Copy + Ord> DamerauLevenshtein<T> {
    pub fn new() -> Self {
        let dists   = RefCell::new(DistMatrix::new(DEFAULT_CAPACITY + 2));
        let last_i1 = RefCell::new(BTreeMap::new());
        Self { dists, last_i1 }
    }

    pub fn distance(&self, slice1: &[T], slice2: &[T]) -> usize {
        let (prefix, postfix) = common_affix_sizes(slice1, slice2);
        let mut slice1 = { let len = slice1.len(); &slice1[prefix .. len - postfix] };
        let mut slice2 = { let len = slice2.len(); &slice2[prefix .. len - postfix] };
        if slice2.len() < slice1.len() {
            std::mem::swap(&mut slice1, &mut slice2);
        }

        let dists = &mut *self.dists.borrow_mut();
        dists.grow(max(slice1.len() + 2, slice2.len() + 2));

        let last_i1 = &mut *self.last_i1.borrow_mut();
        last_i1.clear();

        for (i1, &x1) in slice1.iter().enumerate() {
            let mut l2 = 0;

            for (i2, &x2) in slice2.iter().enumerate() {
                let l1 = *last_i1.get(&x2).unwrap_or(&0);

                unsafe {
                    dists.set(i1 + 2, i2 + 2, min!(
                        dists.get(i1 + 2, i2 + 1) + 1,
                        dists.get(i1 + 1, i2 + 2) + 1,
                        dists.get(i1 + 1, i2 + 1) + (x1 != x2) as usize,
                        dists.get(l1, l2) + (i1 - l1) + (i2 - l2) + 1
                    ));
                }

                if x1 == x2 { l2 = i2 + 1; }
            }
            last_i1.insert(x1, i1 + 1);
        }

        unsafe { dists.get(slice1.len() + 1, slice2.len() + 1) }
    }
}


#[cfg(test)]
mod tests {
    use super::DamerauLevenshtein;

    #[test]
    fn equality() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            vec![],
            vec![1],
            vec![1, 2],
            vec![1, 2, 3],
        ];
        for s in sample.iter() {
            assert_eq!(damlev.distance(s, s), 0);
        }
    }

    #[test]
    fn prefix() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (0, vec![1, 2, 3], vec![1, 2, 3]),
            (1, vec![1, 2, 3], vec![1, 2]),
            (2, vec![1, 2, 3], vec![1]),
            (3, vec![1, 2, 3], vec![]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(damlev.distance(s1, s2), *d);
            assert_eq!(damlev.distance(s2, s1), *d);
        }
    }

    #[test]
    fn add_del_continuous() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1, vec![1, 2, 3], vec![0, 1, 2, 3]),
            (2, vec![1, 2, 3], vec![0, 0, 1, 2, 3]),
            (3, vec![1, 2, 3], vec![0, 0, 0, 1, 2, 3]),

            (1, vec![1, 2, 3], vec![1, 0, 2, 3]),
            (2, vec![1, 2, 3], vec![1, 0, 0, 2, 3]),
            (3, vec![1, 2, 3], vec![1, 0, 0, 0, 2, 3]),

            (1, vec![1, 2, 3], vec![1, 2, 3, 0]),
            (2, vec![1, 2, 3], vec![1, 2, 3, 0, 0]),
            (3, vec![1, 2, 3], vec![1, 2, 3, 0, 0, 0]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(damlev.distance(s1, s2), *d);
            assert_eq!(damlev.distance(s2, s1), *d);
        }
    }

    #[test]
    fn sub_continuous() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1, vec![1, 2, 3, 4], vec![0, 2, 3, 4]),
            (2, vec![1, 2, 3, 4], vec![0, 0, 3, 4]),
            (3, vec![1, 2, 3, 4], vec![0, 0, 0, 4]),

            (1, vec![1, 2, 3, 4], vec![1, 0, 3, 4]),
            (2, vec![1, 2, 3, 4], vec![1, 0, 0, 4]),

            (1, vec![1, 2, 3, 4], vec![1, 2, 3, 0]),
            (2, vec![1, 2, 3, 4], vec![1, 2, 0, 0]),
            (3, vec![1, 2, 3, 4], vec![1, 0, 0, 0]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(damlev.distance(s1, s2), *d);
        }
    }

    #[test]
    fn trans_continuous() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1, vec![1, 2, 3, 4], vec![2, 1, 3, 4]), // swap 1 and 2
            (2, vec![1, 2, 3, 4], vec![2, 1, 4, 3]), // swap 3 and 4
            (3, vec![1, 2, 3, 4], vec![2, 4, 1, 3]), // swap 1 and 4
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(damlev.distance(s1, s2), *d);
        }
    }

    #[test]
    fn add_del_intermittent() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1, vec![1, 2, 3], vec![0, 1, 2, 3]),
            (2, vec![1, 2, 3], vec![0, 1, 0, 2, 3]),
            (3, vec![1, 2, 3], vec![0, 1, 0, 2, 0, 3]),

            (1, vec![1, 2, 3], vec![1, 2, 3, 0]),
            (2, vec![1, 2, 3], vec![1, 2, 0, 3, 0]),
            (3, vec![1, 2, 3], vec![1, 0, 2, 0, 3, 0]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(damlev.distance(s1, s2), *d);
            assert_eq!(damlev.distance(s2, s1), *d);
        }
    }

    #[test]
    fn sub_intermittent() {
        let damlev = DamerauLevenshtein::new();
        let sample = [
            (1, vec![1, 2, 3, 4], vec![0, 2, 3, 4]),
            (2, vec![1, 2, 3, 4], vec![0, 2, 0, 4]),

            (1, vec![1, 2, 3, 4], vec![1, 2, 3, 0]),
            (2, vec![1, 2, 3, 4], vec![1, 0, 3, 0]),
        ];
        for (d, s1, s2) in sample.iter() {
            assert_eq!(damlev.distance(s1, s2), *d);
        }
    }

    #[test]
    fn growth() {
        let damlev = DamerauLevenshtein::new();
        for len in (1..501).step_by(100) {
            let mut v1 = Vec::with_capacity(len);
            let mut v2 = Vec::with_capacity(len);
            v1.resize(len, 1);
            v2.resize(len, 2);
            assert_eq!(damlev.distance(&v1, &v1), 0);
            assert_eq!(damlev.distance(&v1, &[]), len);
            assert_eq!(damlev.distance(&v1, &v2), len);
        }
    }
}
