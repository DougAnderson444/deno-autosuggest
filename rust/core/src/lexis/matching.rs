use std::fmt;
use std::collections::HashSet;
use crate::damlev::DamerauLevenshtein;
use crate::jaccard::Jaccard;
use super::{Word, Text};


const LENGTH_THRESHOLD:  f64 = 0.51;
const JACCARD_THRESHOLD: f64 = 0.41;
const DAMLEV_THRESHOLD:  f64 = 0.21;

thread_local! {
    static DAMLEV:  DamerauLevenshtein<char> = DamerauLevenshtein::new();
    static JACCARD: Jaccard<char>            = Jaccard::new();
}


#[derive(Clone, PartialEq)]
pub struct WordMatch {
    pub query:  MatchSide,
    pub record: MatchSide,
    pub typos:  usize,
    pub fin:    bool,
}


#[derive(Clone, PartialEq, Debug)]
pub struct MatchSide {
    pub ix:      usize,
    pub len:     usize,
    pub slice:   (usize, usize),
    pub primary: bool,
}


impl MatchSide {
    pub fn new(len: usize, slice: (usize, usize), primary: bool) -> Self {
        MatchSide { ix: 0, len, slice, primary }
    }
}


impl fmt::Debug for WordMatch {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "WordMatch {{ ")?;

        for i in 0 .. self.record.len {
            if i == self.record.slice.0 { write!(f, "[")?; }
            write!(f, "r")?;
            if i + 1 == self.record.slice.1 { write!(f, "]")?; }
        }

        write!(f, " /{}/ ", self.typos)?;

        for i in 0 .. self.query.len {
            if i == self.query.slice.0 { write!(f, "[")?; }
            write!(f, "q")?;
            if i + 1 == self.query.slice.1 { write!(f, "]")?; }
        }
        if !self.fin {
            write!(f, "..")?;
        }

        write!(f, " }}")?;
        Ok(())
    }
}


pub fn text_match(rtext: &Text<&[char]>, qtext: &Text<&[char]>) -> Vec<WordMatch> {
    let capacity = min!(rtext.words.len(), qtext.words.len());
    let mut taken: HashSet<usize> = HashSet::with_capacity(capacity);
    let mut matches: Vec<WordMatch> = Vec::with_capacity(capacity);

    for (i, qword) in qtext.words.iter().enumerate() {
        let mut found: Option<WordMatch> = None;

        for (j, rword) in rtext.words.iter().enumerate() {
            if taken.contains(&j) { continue; }
            if let Some(mut m) = word_match(rword, qword, &rtext.chars, &qtext.chars) {
                m.query.ix  = i;
                m.record.ix = j;
                if found.is_none() && !m.record.primary {
                    found = Some(m);
                    continue;
                }
                if m.record.primary {
                    found = Some(m);
                    break;
                }
            }
        }

        if let Some(m) = found {
            taken.insert(m.record.ix);
            matches.push(m);
        }
    }

    matches
}


pub fn word_match(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> Option<WordMatch> {
    if qword.is_empty() || rword.is_empty() {
        return None;
    }
    if !length_check(rword, qword) {
        return None;
    }
    if !jaccard_check(rword, qword, rchars, qchars) {
        return None;
    }

    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let mut best_match = None;

    DAMLEV.with(|damlev| {
        damlev.distance(qchars, rchars);
        let dists  = &*damlev.dists.borrow();

        let left  = if qword.fin { max!(qword.stem, rword.stem) } else { qword.stem } - 1;
        let right = max!(qword.len(), rword.len()) + 1;

        if right <= left { return; }

        let range = (left .. right).rev();

        for rlen in range.clone() {
            for qlen in range.clone() {
                // Out of bounds.
                if qlen > qword.len() { continue; }
                if rlen > rword.len() { continue; }
                // Left margin is for insertion/deletion, not for both prefixes at the same time.
                if rlen == left  && qlen == left  { continue; }
                // Compare full words only if query is finished.
                if qword.fin && rlen < rword.stem { break; }
                if qword.fin && qlen < qword.stem { break; }
                // Words with 2+ insertions/deletions are mismatched by default.
                if (qlen as isize - rlen as isize).abs() > 1 { continue; }

                let dist = dists.get(qlen + 1, rlen + 1);
                let rel  = dist as f64 / max!(qlen, rlen, 1) as f64;

                if rel > DAMLEV_THRESHOLD { continue; }

                match best_match {
                    None => {
                        best_match = Some(WordMatch {
                            query:  MatchSide::new(qlen,        (0, qlen), qword.is_primary()),
                            record: MatchSide::new(rword.len(), (0, rlen), rword.is_primary()),
                            typos:  dist,
                            fin:    qword.fin || rword.len() == rlen,
                        });
                    },
                    Some(ref mut m) => {
                        if m.typos <= dist { continue; }
                        m.record.slice = (0, rlen);
                        m.typos        = dist;
                        m.fin          = qword.fin || rword.len() == rlen;
                    },
                }

                if dist == 0 {
                    break;
                }
            }
        }
    });

    best_match
}


pub fn length_check(rword: &Word, qword: &Word) -> bool {
    let qlen  = qword.len();
    let rlen  = if qword.fin { rword.len() } else { min!(qlen, rword.len()) };

    if qlen <= 3 || rlen <= 3 {
        return qlen == rlen;
    }

    let long  = max!(qlen, rlen);
    let short = min!(qlen, rlen);
    let dist  = 1.0 - (short as f64 / long as f64);

    dist < LENGTH_THRESHOLD
}


pub fn jaccard_check(rword: &Word, qword: &Word, rchars: &[char], qchars: &[char]) -> bool {
    let rchars = rword.view(rchars);
    let qchars = qword.view(qchars);
    let rslice = if qword.fin { &rchars } else { &rchars[.. min!(qword.len(), rword.len())] };
    let dist   = JACCARD.with(|j| j.rel_dist(&rslice, &qchars));
    dist < JACCARD_THRESHOLD
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lexis::Chars;
    use crate::lang::lang_english;
    use super::{Word, Text, word_match, text_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&Chars::Whitespaces)
    }


    // Match word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_empty_both() {
        let qchars = "".chars().collect::<Vec<_>>();
        let rchars = "".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_record() {
        let qchars = "mailbox".chars().collect::<Vec<_>>();
        let rchars = "".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    #[test]
    fn match_word_empty_query() {
        let qchars = "".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_eq!(word_match(&rword, &qword, &rchars[..], &qchars[..]), None);
    }


    // Match word: prefixed by same length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_full_strict() {
        let qchars = "mailbox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_insertion() {
        let qchars = "mailybox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_deletion() {
        let qchars = "mailox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_fuzzy_transposition() {
        let qchars = "maiblox".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_full_query_too_long() {
        let qchars1 = "mailboxes".chars().collect::<Vec<_>>();
        let qchars2 = "mailboxes".chars().collect::<Vec<_>>();
        let rchars  = "mail"     .chars().collect::<Vec<_>>();
        let qword1  = Word::new(qchars1.len()).fin(true);
        let qword2  = Word::new(qchars2.len()).fin(false);
        let rword   = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }

    #[test]
    fn match_word_full_stem() {
        let     rchars  = "universe"  .chars().collect::<Vec<_>>();
        let     qchars1 = "university".chars().collect::<Vec<_>>();
        let     qchars2 = "university".chars().collect::<Vec<_>>();
        let mut rword   = Word::new(rchars.len());
        let mut qword1  = Word::new(qchars1.len());
        let     qword2  = Word::new(qchars2.len());

        let lang = lang_english();
        qword1.stem(&qchars1[..], &lang);
        rword .stem(&rchars[..],  &lang);

        assert_debug_snapshot!(word_match(&rword, &qword1, &rchars[..], &qchars1[..]));
        assert_debug_snapshot!(word_match(&rword, &qword2, &rchars[..], &qchars2[..]));
    }


    // Match word: prefixed by lesser length word
    // ----------------------------------------------------------------

    #[test]
    fn match_word_partial_strict() {
        let qchars = "mailb".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword = Word::new(qchars.len()).fin(false);
        let rword = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_insertion() {
        let qchars = "maiylb".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_deletion() {
        let qchars = "maib".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    #[test]
    fn match_word_partial_fuzzy_transposition() {
        let qchars = "malib".chars().collect::<Vec<_>>();
        let rchars = "mailbox".chars().collect::<Vec<_>>();
        let qword  = Word::new(qchars.len()).fin(false);
        let rword  = Word::new(rchars.len());
        assert_debug_snapshot!(word_match(&rword, &qword, &rchars[..], &qchars[..]));
    }


    // Match text
    // ----------------------------------------------------------------

    #[test]
    fn match_text_empty_both() {
        let qtext = text("").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_empty_one() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&qtext.to_ref(), &rtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_equality() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_typos() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_first() {
        let qtext = text("yelow").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_second() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_unfinished() {
        let qtext = text("maiblox yel").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_intersection() {
        let qtext = text("big malibox yelo").fin(false);
        let rtext = text("small yellow metal mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword() {
        let mut qtext = text("the").fin(false);
        let mut rtext = text("the theme");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));

        let lang = lang_english();
        qtext = qtext.mark_pos(&lang);
        rtext = rtext.mark_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_regression_best_match() {
        let qtext = text("sneak").fin(false);
        let rtext = text("sneaky");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }
}