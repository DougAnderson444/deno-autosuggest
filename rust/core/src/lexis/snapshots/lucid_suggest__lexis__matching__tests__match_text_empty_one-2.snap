---
source: src/lexis/matching.rs
expression: q.matches(&r)
---
[]
